plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -103,58 +54,51 @@ not have side effects; (2) all arguments be pure, since otherwise it is unsafe t
-it's `Utils.acceptable`.
--}
-splitParams ::
-  Term tyname name uni fun ann ->
-  (Arity tyname name, Term tyname name uni fun ann)
-splitParams = \case
-  LamAbs _ n _ body ->
-    let (nextArgs, nextBody) = splitParams body in (TermParam n : nextArgs, nextBody)
-  TyAbs _ n _ body ->
-    let (nextArgs, nextBody) = splitParams body in (TypeParam n : nextArgs, nextBody)
-  -- Whenever we encounter a body that is not a lambda or type abstraction, we are done counting
-  tm -> ([], tm)
-
-{- | Fully apply the RHS of the given variable to the given arguments, and beta-reduce
+{- | Apply the RHS of the given variable to the given arguments, and beta-reduce
 the application, if possible.
 -}
-fullyApplyAndBetaReduce ::
+applyAndBetaReduce ::
   forall tyname name uni fun ann.
   (InliningConstraints tyname name uni fun) =>
-  -- | The variable
-  VarInfo tyname name uni fun ann ->
+  -- | The rhs of the variable, should have been renamed already
+  Term tyname name uni fun ann ->
   -- | The arguments
   AppContext tyname name uni fun ann ->
   InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
-fullyApplyAndBetaReduce info args0 = do
-  rhsBody <- liftDupable (let Done rhsBody = varRhsBody info in rhsBody)
-  let go ::
+applyAndBetaReduce rhs args0 = do
+  let -- | Drop one term or type lambda abstraction of the given term.
+      getFnBody :: Term tyname name uni fun ann -> Term tyname name uni fun ann
+      getFnBody (LamAbs _ann _n _ty body) = body
+      getFnBody (TyAbs _ann _n _kd body)  = body
+      getFnBody tm                        = tm
+      go ::
+        -- | The rhs of the variable, should have been renamed already
         Term tyname name uni fun ann ->
-        Arity tyname name ->
         AppContext tyname name uni fun ann ->
         InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
-      go acc arity args = case (arity, args) of
-        -- success
-        ([], _) -> pure . Just $ fillAppContext acc args
-        (TermParam param : arity', TermAppContext arg _ args') -> do
-          safe <- safeToBetaReduce param arg
-          if safe
+      go acc args = case (acc, args) of
+        (LamAbs _ann n _ty _tm, TermAppContext arg _ args') -> do
+          safe <- safeToBetaReduce n arg acc
+          if safe -- we only do substitution if it is safe to beta reduce
             then do
-              acc' <-
-                termSubstNamesM
-                  (\n -> if n == param then Just <$> PLC.rename arg else pure Nothing)
-                  acc
-              go acc' arity' args'
+              acc' <- do
+                termSubstNamesM -- substitute the term param with the arg in the function body
+                  -- rename before substitution to ensure global uniqueness
+                  (\tmName -> if tmName == n then Just <$> PLC.rename arg else pure Nothing)
+                  (getFnBody acc) -- drop the beta reduced term lambda
```

Comment:
Returning `tm` would be wrong I think, because then nothing will ever be substituted! We need the `acc` with one lambda dropped to get the processed part.

Reply:
You're doing `case` precisely on `acc`, no?

Reply:
Oh you're right this is the input actually.